addons/pos_cache/models/pos_cache.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import cPickle
+
+from openerp import models, fields, api
+from openerp.exceptions import UserError
+from openerp.tools.safe_eval import safe_eval as eval
+
+
+class pos_cache(models.Model):
+    _name = 'pos.cache'
+
+    cache = fields.Binary()
+    product_domain = fields.Text(required=True)
+    product_fields = fields.Text(required=True)
+
+    config_ids = fields.One2many('pos.config', 'cache_id', 'Config')
+    compute_user_id = fields.Many2one('res.users', 'Cache compute user',
+                                      help="User used to compute the cache content and thus avoid the bad practice of using admin in multi-company")
+
+    @api.model
+    def refresh_all_caches(self):
+        self.env['pos.config'].search([('state', '=', 'active')]).recompute_cache()
+
+    @api.one
+    def refresh_cache(self):
+        products = self.env['product.product'].search(eval(self.product_domain))
+        prod_ctx = products.with_context(pricelist=self.config_ids[0].pricelist_id.id, display_default_code=False)
+        if self.compute_user_id:
+            prod_ctx = prod_ctx.sudo(self.compute_user_id.id)
+        else:
+            self.compute_user_id = self.env.uid
+        res = prod_ctx.read(eval(self.product_fields))
+        datas = {
+            'cache': cPickle.dumps(res, protocol=cPickle.HIGHEST_PROTOCOL),
+        }
+
+        self.write(datas)
+
+    @api.model
+    def get_cache(self):
+        return cPickle.loads(self.cache)
+
+
+class pos_config(models.Model):
+    _inherit = 'pos.config'
+
+    # Use a related model to avoid the load of the cache when the pos load his config
+    cache_id = fields.Many2one('pos.cache', 'Cache')
+
+    cache_time = fields.Datetime(related='cache_id.write_date', string='Last cache compute', readonly=True)
+    cache_compute_user_id = fields.Many2one('res.users', related='cache_id.compute_user_id')
+
+    @api.multi
+    def unlink(self):
+        cache_ids = [config.cache_id for config in self]
+        to_return = super(pos_config, self).unlink()
+
+        for cache_id in cache_ids:
+            cache_id.unlink()
+
+        return to_return
+
+    def _js_domain_to_python_domain(self, js_domain):
+        python_domain = [(criteria[0], criteria[1], criteria[2]) for criteria in js_domain]
+        return python_domain
+
+    @api.multi
+    def get_products_from_cache(self, fields, domain):
+        if not self.cache_id:
+            self.recompute_cache(fields, domain)
+        return self.cache_id.get_cache()
+
+    @api.one
+    def recompute_cache(self, fields=[], domain=[]):
```

Comment:
Please use a non mutable value like None, (even if there are no bug here) it's very touchy.